src/FormBuilder/Controllers/FormBuilderResourceController.php

```diff
@@ -24,50 +24,89 @@ public function show(WP_REST_Request $request)
     {
         $formId = $request->get_param('id');
 
-        if (!get_post($formId)) {
+        /** @var DonationForm $form */
+        $form = DonationForm::find($formId);
+
+        if (!$form) {
             return rest_ensure_response(new WP_Error(404, 'Form not found.'));
         }
 
-        $formData = get_post($formId)->post_content;
-        $formBuilderSettings = get_post_meta($formId, 'formBuilderSettings', true);
+        if ($blockError = $this->validateBlocks($form->blocks)) {
+            return rest_ensure_response($blockError);
+        }
 
         return rest_ensure_response([
-            'blocks' => $formData,
-            'settings' => $formBuilderSettings
+            'blocks' => $form->blocks->toJson(),
+            'settings' => json_encode($form->settings)
         ]);
     }
 
     /**
      * Update the form builder
      *
-     * TODO: replace logic with form model
-     * TODO: handle more validation and errors
-     *
      * @unreleased
      *
      * @return WP_Error|WP_HTTP_Response|WP_REST_Response
+     * @throws Exception
      */
     public function update(WP_REST_Request $request)
     {
         $formId = $request->get_param('id');
         $formBuilderSettings = $request->get_param('settings');
-        $data = $request->get_param('blocks');
+        $rawBlocks = $request->get_param('blocks');
 
-        if (!get_post($formId)) {
-            return rest_ensure_response(new WP_Error(404, 'Form not found.'));
+        /** @var DonationForm $form */
+        $form = DonationForm::find($formId);
+
+        if (!$form) {
+            return rest_ensure_response(new WP_Error(404, __('Form not found.', 'give')));
         }
 
-        $meta = update_post_meta($formId, 'formBuilderSettings', $formBuilderSettings);
+        $blocks = BlockCollection::fromJson($rawBlocks);
 
-        $post = wp_update_post([
-            'ID' => $formId,
-            'post_content' => $data,
-            'post_title' => json_decode($formBuilderSettings, false)->formTitle,
-        ]);
+        if ($blockError = $this->validateBlocks($blocks)) {
+            return rest_ensure_response($blockError);
+        }
+
+        $updatedSettings = json_decode($formBuilderSettings, true);
+        $form->settings = array_merge($form->settings ?? [], $updatedSettings);
+        $form->title = $updatedSettings['formTitle'];
+        $form->blocks = $blocks;
+        $form->save();
 
         return rest_ensure_response([
-            'settings' => $meta,
-            'form' => $post,
+            'settings' => json_encode($form->settings),
+            'form' => $form->id,
         ]);
     }
+
+    /**
+     * @unreleased
+     *
+     * @return string[]
+     */
+    protected function getRequiredBlockNames(): array
+    {
+        return [
+            "custom-block-editor/donation-amount",
+            "custom-block-editor/donor-info",
```

Comment:
The donor-info block contains the name & email fields as inner blocks

Reply:
@jonwaldstein if you haven't already started on this, then I'll take it.

Reply:
@kjohnson go for it, i'm not that familiar with how those are setup right now.

Reply:
@JasonTheAdams I replaced the specific section blocks with a single section block that is re-used in the default form configuration.

@jonwaldstein I also updated the required validation in the `FormBuilderResourceController` to check for field's by name in the Schema.

Reply:
@kjohnson nice, mind updating `TestsNextGen\Feature\Controllers\FormBuilderResourceControllerTest::testUpdateShouldFailIfSettingsAreInvalid` as well?

Reply:
Also, I'm seeing amount, name, and gatewayId in the required field names — we also need email.

Reply:
@JasonTheAdams I added the email fields as required, see https://github.com/impress-org/givewp-next-gen/pull/63/commits/9ba631d87773b89322af46c18602f0dbabdb522b.

Reply:
@jonwaldstein I updated the test, see https://github.com/impress-org/givewp-next-gen/pull/63/commits/f8960eda87c01ed049bd7913cd52b850122422f6.

Also note that the refactor away form specific sections required a change to how the hidden fields are added, if you could please take a look at that as well.

Reply:
okay great, thanks

Reply:
@kjohnson i'm getting a NameCollision error from the fields api that is preventing the form to update.

Reply:
@jonwaldstein resolved by https://github.com/impress-org/givewp-next-gen/pull/63/commits/5d2e99fee2f81c0bf8b986b86cbdfba97c56d667. The Fields API requires that node names be unique and since we moved to a single re-usable section block we needed to create a unique name. This is generated on demand, since sections don't have a name setting. Note that fields have still be plucked by name recursively, so the section names aren't necessary for accessing specific fields.